pkg/source/iziswap/pool_simulator.go

```diff
@@ -1,251 +1,172 @@
-	tokenInIndex := p.GetTokenIndex(tokenInAddr)
-	tokenOutIndex := p.GetTokenIndex(tokenOutAddr)
+	tokenIn, tokenOut := tokenAmountIn.Token, param.TokenOut
+	tokenInIndex, tokenOutIndex := p.GetTokenIndex(tokenIn), p.GetTokenIndex(tokenOut)
 	if tokenInIndex < 0 || tokenOutIndex < 0 || tokenInIndex == tokenOutIndex {
-		return &pool.CalcAmountOutResult{}, fmt.Errorf("tokenInIndex %v or tokenOutIndex %v is not correct", tokenInIndex, tokenOutIndex)
+		return nil, ErrInvalidToken
 	}
 
-	// Clone tokenAmountIn.Amount, since the SDK will mutate it
-	tokenAmountInAmount := new(big.Int).Set(tokenAmountIn.Amount)
+	tokenAmountInAmount, overflow := uint256.FromBig(tokenAmountIn.Amount)
+	if overflow {
+		return nil, ErrInvalidAmount
+	}
 
-	x2y := tokenInAddr < tokenOutAddr
+	x2y := tokenIn < tokenOut
+	ptLimit := p.PoolInfo.CurrentPoint // TODO: not limit swap-range in the future or give a way to modify it
+	swapFn := swap.SwapX2Y
 	if x2y {
-		// todo, not limit swap-range in the future
-		//    or give a way to modify it
-		lowPt := p.PoolInfo.CurrentPoint - SIMULATOR_PT_RANGE
-		ret, err := swap.SwapX2Y(tokenAmountInAmount, lowPt, p.PoolInfo)
-		if err != nil {
-			return &pool.CalcAmountOutResult{}, err
-		}
-		amountY := ret.AmountY
-		// // Fee can be ignored for now
-		// amountX := ret.AmountX
-		// fee := new(big.Int).Mul(&amountX, big.NewInt(int64(p.PoolInfo.Fee)))
-		// fee.Div(fee, feeBase)
-		return &pool.CalcAmountOutResult{
-			TokenAmountOut: &pool.TokenAmount{
-				Token:  tokenOut,
-				Amount: amountY,
-			},
-			Fee: &pool.TokenAmount{
-				Token:  tokenAmountIn.Token,
-				Amount: nil,
-			},
-			Gas: p.gas.Swap,
-			SwapInfo: iZiSwapInfo{
-				nextPoint:      ret.CurrentPoint,
-				nextLiquidity:  new(big.Int).Set(ret.Liquidity),
-				nextLiquidityX: new(big.Int).Set(ret.LiquidityX),
-			},
-		}, nil
+		ptLimit -= SIMULATOR_PT_RANGE
 	} else {
-		// todo, not limit swap-range in the future
-		//    or give a way to modify it
-		highPt := p.PoolInfo.CurrentPoint + SIMULATOR_PT_RANGE
-		ret, err := swap.SwapY2X(tokenAmountInAmount, highPt, p.PoolInfo)
-		if err != nil {
-			return &pool.CalcAmountOutResult{}, err
-		}
-		amountX := ret.AmountX
-		// // Fee can be ignored for now
-		// fee := new(big.Int).Mul(amountX, big.NewInt(int64(p.PoolInfo.Fee)))
-		// fee.Div(fee, feeBase)
-		return &pool.CalcAmountOutResult{
-			TokenAmountOut: &pool.TokenAmount{
-				Token:  tokenOut,
-				Amount: amountX,
-			},
-			Fee: &pool.TokenAmount{
-				Token:  tokenAmountIn.Token,
-				Amount: nil,
-			},
-			Gas: p.gas.Swap,
-			SwapInfo: iZiSwapInfo{
-				nextPoint:      ret.CurrentPoint,
-				nextLiquidity:  new(big.Int).Set(ret.Liquidity),
-				nextLiquidityX: new(big.Int).Set(ret.LiquidityX),
-			},
-		}, nil
+		ptLimit += SIMULATOR_PT_RANGE
+		swapFn = swap.SwapY2X
 	}
+
+	ret, err := swapFn(tokenAmountInAmount, ptLimit, p.PoolInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	amountOut := ret.AmountX
+	if x2y {
+		amountOut = ret.AmountY
+	}
```

Comment:
```suggestion
	amountOut := lo.Ternary(x2y, ret.AmountY, ret.AmountX)
```

I believe we don't want to see the condition check anymore